src/main/kotlin/graphics/scenery/CurveGeometry.kt

```diff
@@ -0,0 +1,189 @@
+                        verticesVectors.add(curveGeometry[shapeIndex + 1][vertexIndex + 1])
+                        verticesVectors.add(curveGeometry[shapeIndex + 1][vertexIndex])
+                    }
+                    if(vertexIndex == shape.lastIndex) {
+                        verticesVectors.add(curveGeometry[shapeIndex][0])
+                        verticesVectors.add(curveGeometry[shapeIndex+1][0])
+                        verticesVectors.add(curveGeometry[shapeIndex+1][shape.lastIndex])
+
+                        verticesVectors.add(curveGeometry[shapeIndex+1][shape.lastIndex])
+                        verticesVectors.add(curveGeometry[shapeIndex][shape.lastIndex])
+                        verticesVectors.add(curveGeometry[shapeIndex][0])
+                    }
+                }
+            }
+        }
+        vertices = BufferUtils.allocateFloat(verticesVectors.size*3)
+        verticesVectors.forEach{
+            vertices.put(it.xyz().toFloatArray())
+        }
+        vertices.flip()
+        texcoords = BufferUtils.allocateFloat(verticesVectors.size*2)
+        recalculateNormals()
+    }
+
+    /**
+     * This function calculates the tangent at a given index in the catmull rom curve.
+     * @param [i] index of the curve (not the geometry!)
+     */
+    fun getTangent(i: Int): GLVector {
+        val s = cur.size
+        return when(i) {
+            0 -> ((cur[i+1] - cur[i]).normalized)
+            (s-2) -> ((cur[i+1] - cur[i]).normalized)
+            (s-1) -> ((cur[i] - cur[i-1]).normalized)
+            else -> ((cur[i+1] - cur[i-1]).normalized)
+        }
+    }
+
+
+    /**
+     * This function returns the frenet frames along the curve. This is essentially a new
+     * coordinate system which represents the form of the curve. For details concerning the
+     * calculation see: http://www.cs.indiana.edu/pub/techreports/TR425.pdf
+     */
+    data class FrenetFrame(val tangent: GLVector, var normal: GLVector?, var bitangent: GLVector?, val translation: GLVector)
+    fun computeFrenetFrames(): List<FrenetFrame> {
+
+        val frenetFrameList = ArrayList<FrenetFrame>(cur.size)
+
+        //adds all the tangent vectors
+        for(i in 0 until cur.size) {
+            val frenetFrame = FrenetFrame(getTangent(i), null, null, cur[i])
+            frenetFrameList.add(frenetFrame)
+        }
+
+        //initial normal vector perpendicular to first tangent vector
+        var vec: GLVector
+        vec = if(frenetFrameList[0].tangent.x() >= 0.9f || frenetFrameList[0].tangent.z() >= 0.9f) {
+            GLVector(0f, 1f, 0f)
+        }
+        else {
+            GLVector(1f, 0f, 0f)
+        }
+
+        val normal = frenetFrameList[0].tangent.cross(vec)
+
+        frenetFrameList[0].normal = normal
+        frenetFrameList[0].bitangent = frenetFrameList[0].tangent.cross(normal)
+
+        for(i in 0 until (frenetFrameList.size-1)) {
```

Comment:
Instead of indexing, use `frenetFrameList.windowed(2, 1).forEach { (firstFrame, secondFrame) -> ...`